### PR TITLE
Add LDAP authentication to pgbouncer.

### PR DESCRIPTION
**Pull request by zhaorui20005**
Since the only way to support ldap is pam in pgbouncer and configuring pam with ldap is quite hard to succeed. So This commit add ldap support to allow more users authenticate with ldap in pgbouncer.

And because some of the ldap library functions are run in synchronization way, so we follow the thread model of PAM which has already been adopted.

The following configs should be set to enable ldap authentication:
1. auth_file
    Example:
    `cat user.txt`
    "user1"  "" ; We sugguest an empty password

2. auth_hba_file
    Example:
    `cat hba.conf`
    `host all user1 0.0.0.0/0 ldap ldapserver=10.117.190.215 ldapbasedn="CN=Users,DC=greenplum,DC=org" ldapbinddn="CN=Administrator,CN=Users,DC=greenplum,DC=org" ldapbindpasswd="Welcome1!" ldapsearchattribute="sAMAccountname"`

3. auth_type
    `auth_type = hba`

And if you have enabled ldap authentication between psql and pgbouncer,
    you should set the authentication method in postgres as trust so that
    users could change the password in OPENLDAP or AD server as they want.
    Also you can set the method to md5/password/scram-sha-256 if you have
    entered a password and you don't need to change the password.

Example to start pgbouncer with ldap to auth user1:

`./pgbouncer -d pgbouncer.conf`

```
cat pgbouncer.conf 
[databases]
* = port = 6000 host=127.0.0.1

[pgbouncer]
listen_addr = 0.0.0.0
listen_port = 6432

auth_type = hba
auth_file = user.txt
auth_hba_file = hba.conf

cat user.txt
"user1" ""

cat hba.conf
host all user1 0.0.0.0/0 ldap ldapserver=127.0.0.1 ldapbasedn="ou=people,dc=zdpoc,dc=com" ldapbinddn="cn=admin,dc=zdpoc,dc=com" ldapbindpasswd="1"
```

### CODE REVIEW

**Review by lij55 — APPROVED** [review]

File: src/hba.c [hunk 1]
```diff
@@ -743,3 +771,49 @@ int hba_eval(struct HBA *hba, PgAddr *addr, bool is_tls, const char *dbname, con
 	}
 	return AUTH_REJECT;
 }
+#ifdef HAVE_LDAP
+char *get_hba_complexline(struct HBA *hba, PgAddr *addr, bool is_tls, const char *dbname, const char *username)
```
Comment (kensou97) [inline src/hba.c:775]: Is it possible to change hba_eval() return HBARule* and remove this method? They look like duplicate code.

File: src/hba.c [hunk 2]
```diff
@@ -51,6 +51,9 @@ struct HBARule {
 	uint8_t rule_mask[16];
 	struct HBAName db_name;
 	struct HBAName user_name;
+#ifdef HAVE_LDAP
+	char *line_after_method;
```
Comment (kensou97) [inline src/hba.c:55]: The variable name is a little confusing at first glance...It's called "auth-options" [here](https://www.postgresql.org/docs/current/auth-pg-hba-conf.html), will "auth_options" or "rule_options" be a better name?

File: src/hba.c (hunk 1 above)
Comment (kensou97) [inline src/hba.c:775]: Is it possible to change hba_eval() return HBARule* and remove this method? They look like duplicate code.

File: src/hba.c (hunk 2 above)
Comment (zhaorui20005) [reply src/hba.c]: I think "auth_options" is a better name , thx!

File: src/hba.c (hunk 1 above)
Comment (zhaorui20005) [reply src/hba.c]: I have pushed another commit to fix this, please help to review, thank you for your comments.

File: configure.ac [hunk 3]
```diff
@@ -79,6 +79,28 @@ AC_ARG_WITH(pam,
...
+  [ LDAP=
+    if test "$withval" != no; then
+        have_pthreads=no
+        # Look for LDAP header and lib
+        AC_CHECK_HEADERS(ldap.h, [have_ldap_header=t])
+        AC_CHECK_HEADERS(pthread.h, [have_pthreads=yes])
+        AC_SEARCH_LIBS(ldap_initialize, ldap, [have_libldap=t])
+        AC_SEARCH_LIBS(pthread_create, pthread, [], [have_pthreads=no])
+        if test x"${have_pthreads}" != xyes; then
+           AC_MSG_ERROR([pthread library should be available for PAM support])
```
Comment (huiliang-liu) [inline configure.ac:95]: PAM --> LDAP

Comment (zhaorui20005) [reply configure.ac]: Fixed!

**Pull request merged**